cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -132,4 +179,88 @@ public void create() throws IOException {
       LOG.info("Upgraded hbase table: {}", tableName);
     }
   }
+
+
+  // Moves dataset instance meta entries into new table (in system namespace)
+  // Also updates the spec's name ('cdap.user.foo' -> 'foo')
+  private void upgradeDatasetInstanceMDS() throws IOException, DatasetManagementException, InterruptedException,
+    TransactionFailureException {
+    LOG.info("Upgrading dataset instance mds.");
+    datasetInstanceMds.execute(new TransactionExecutor.Function<UpgradeMdsStores<DatasetInstanceMDS>, Void>() {
+      @Override
+      public Void apply(UpgradeMdsStores<DatasetInstanceMDS> ctx) throws Exception {
+        MDSKey key = new MDSKey(Bytes.toBytes(DatasetInstanceMDS.INSTANCE_PREFIX));
+        DatasetInstanceMDS newMds = ctx.getNewMds();
+        Map<MDSKey, DatasetSpecification> dsSpecs = ctx.getOldMds().listKV(key, DatasetSpecification.class);
+        for (Map.Entry<MDSKey, DatasetSpecification> dsSpecEntry : dsSpecs.entrySet()) {
+          DatasetSpecification dsSpec = dsSpecEntry.getValue();
+          LOG.info("Migrating dataset Spec: {}", dsSpec);
+          Id.Namespace namespace = namespaceFromDatasetName(dsSpec.getName());
+          DatasetSpecification migratedDsSpec = migrateDatasetSpec(dsSpec);
+          LOG.info("Writing new dataset Spec: {}", migratedDsSpec);
+          newMds.write(namespace, migratedDsSpec);
+        }
+        return null;
+      }
+    });
+  }
+
+  /**
+   * Construct a {@link TableId} from a dataset name
+   * TODO: CDAP-1593 - This is bad and should be removed, since it makes assumptions about the dataset name format.
+   *
+   * @param name the dataset/table name to construct the {@link TableId} from
+   * @return the {@link TableId} object for the specified dataset/table name
+   */
+  private static TableId from(String name) {
```

Comment:
My concern again - We're using a *table* abstraction, that is really meant for use in the storage providers for an operation on a *dataset*. It will create a problem in future, if our dataset name has absolutely no relation to table name.  Unlikely, but the assumption that ``dataset name == table name`` has caused problems in the last couple of weeks.